### ISSUE TITLE
CustomImportOrder check: THIRD_PARTY_PACKAGE should exclude SAME_PACKAGE(n) imports

### ISSUE THREAD

**Issue by ivanov-alex** [checkstyle/checkstyle#1273]
According to documentation THIRD_PARTY_PACKAGE imports should exclude SAME_PACKAGE(n) imports: `3) THIRD_PARTY_PACKAGE group. This group sets ordering of third party imports. Third party imports are all imports except STATIC, SAME_PACKAGE(n), STANDARD_JAVA_PACKAGE and SPECIAL_IMPORTS. `

However, in fact they are not. If THIRD_PARTY_PACKAGE preceeds SAME_PACKAGE on a list of rules and SAME_PACKAGE import matches RegExp for THIRD_PARTY_PACKAGE group (default value is ".*" which will cover all the imports), it will be included into THIRD_PARTY_PACKAGE.
For example:
```java
package com.puppycrawl.tools.checkstyle.imports;

import com.google.common.annotations.GwtCompatible;
import com.puppycrawl.tools.*;
import com.google.common.annotations.Beta;
```
If configuration has THIRD_PARTY_PACKAGE before SAME_PACKAGE
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
<module name="CustomImportOrder">
    <property name="customImportOrderRules"
    value="THIRD_PARTY_PACKAGE###SAME_PACKAGE(3)###SPECIAL_IMPORTS"/>
    <property name="specialImportsRegExp" value="android.*"/>
</module>
    </module>
</module>
```
Output will have no errors (all 3 imports will be treated as one group: THIRD_PARTY_PACKAGE)
```
Starting audit...
Audit done.
```
If configuration has SAME_PACKAGE before THIRD_PARTY_PACKAGE
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
<module name="CustomImportOrder">
    <property name="customImportOrderRules"
    value="SAME_PACKAGE(3)###THIRD_PARTY_PACKAGE###SPECIAL_IMPORTS"/>
    <property name="specialImportsRegExp" value="android.*"/>
</module>
    </module>
</module>
```
Output will show error for  `import com.puppycrawl.tools.*;`
```
Starting audit...
/home/alex/test.java:4: warning: Import statement is in the wrong order. Should be in the 'SAME_PACKAGE' group.
Audit done.
```

**Comment by ivanov-alex** [comment]
It appears that CustomImportOrder handles groups in a pretty strange way.
For example, static imports can be assigned only to STATIC or NON_GROUP. They cannot be assigned to SAME_PACKAGE, STANDARD, THIRD_PARTY or SPECIAL groups.

Example with SPECIAL:
```java
1 package com.github.sevntu.checkstyle.checks.imports;
2 
3 import static org.junit.Assert.fail;
4 import org.junit.Test;
```
```xml
        <module name="CustomImportOrder">
            <property name="customImportOrderRules" value="SPECIAL_IMPORTS"/>
            <property name="specialImportsRegExp" value="org"/>
            <property name="sortImportsInGroupAlphabetically" value="true"/>
	        <property name="separateLineBetweenGroups" value="true"/>
        </module>
```
`test.java:4: Import statement is in the wrong order. Should be in the 'SPECIAL_IMPORTS' group.`

In a similar way, there is a [UT](https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheckTest.java#L125) that has only `STANDARD_JAVA_PACKAGE` on configuration, all the first imports [static lines 3-5 and non-static lines 7-11](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/imports/InputCustomImportOrder2.java) are from standard packages, but UT expects errors on lines 7-11 claiming that these imports should be in a STANDARD group.

I propose establishing priorities for the groups. In situation, where one import can be assigned to several groups, it will be assigned to group with top priority.
I suggest following priorities:

1. STATIC has top priority
2. SAME_PACKAGE, STANDARD and SPECIAL have priority according to their order on configuration. The earlier, the higher.
3. THIRD_PARTY has the least priority

**Comment by ivanov-alex** [comment]
@romani Could you please take a look at my comment above? Is it acceptable? Maybe you have better idea or different vision.

**Comment by romani** [comment]
Logic is correct, please proceed witg implementation

**Comment by romani** [comment]
please do not forget to put this examples to javadoc/xdoc - this will help a lot to user to understand a rule.

**Comment by romani** [comment]
merged
